src/shapes/sdfgrid.cpp

```diff
@@ -562,70 +961,95 @@ class SDFGrid final : public SDF<Float, Spectrum> {
+                        (x + 1) * shape_rcp[2], (y + 0) * shape_rcp[1],
+                        (z + 1) * shape_rcp[0])));
                     bbox.expand(to_world.transform_affine(ScalarPoint3f(
-                        (x + 0) * shape_rcp[2], (y + 1) * shape_rcp[1], (z + 1) * shape_rcp[0])));
+                        (x + 0) * shape_rcp[2], (y + 1) * shape_rcp[1],
+                        (z + 1) * shape_rcp[0])));
                     bbox.expand(to_world.transform_affine(ScalarPoint3f(
-                        (x + 1) * shape_rcp[2], (y + 1) * shape_rcp[1], (z + 1) * shape_rcp[0])));
+                        (x + 1) * shape_rcp[2], (y + 1) * shape_rcp[1],
+                        (z + 1) * shape_rcp[0])));
 
                     size_t voxel_index =
                         (x + 0) + (y + 0) * (shape_v[0] - 1) +
                         (z + 0) * (shape_v[0] - 1) * (shape_v[1] - 1);
+
                     voxel_indices[count] = voxel_index;
-                    aabbs[count] = optix::BoundingBox3f(bbox);
+                    aabbs[count]         = BoundingBoxType(bbox);
 
                     count++;
                 }
             }
         }
 
-        jit_free(grid);
+        void *aabbs_ptr           = nullptr;
+        size_t *voxel_indices_ptr = nullptr;
+
+        if constexpr (dr::is_cuda_v<Float>) {
+            // TODO: async memcpy
+            aabbs_ptr =
+                jit_malloc(AllocType::Device, sizeof(BoundingBoxType) * count);
+            jit_memcpy(JitBackend::CUDA, aabbs_ptr, aabbs,
+                       sizeof(BoundingBoxType) * count);
 
-        //TODO: async memcpy
-        void *aabbs_ptr = jit_malloc(AllocType::Device, sizeof(optix::BoundingBox3f) * count);
-        jit_memcpy(JitBackend::CUDA, aabbs_ptr, aabbs, sizeof(optix::BoundingBox3f) * count);
+            voxel_indices_ptr = (size_t *) jit_malloc(AllocType::Device,
+                                                      sizeof(size_t) * count);
+            jit_memcpy(JitBackend::CUDA, voxel_indices_ptr, voxel_indices,
+                       sizeof(size_t) * count);
 
-        void *voxel_indices_ptr = jit_malloc(AllocType::Device, sizeof(size_t) * count);
-        jit_memcpy(JitBackend::CUDA, voxel_indices_ptr, voxel_indices, sizeof(size_t) * count);
+            jit_free(aabbs);
+            jit_free(voxel_indices);
+            jit_free(grid);
 
-        return {aabbs_ptr, voxel_indices_ptr, count};
+        } else {
+            // Our initial allocations were host-visible, so just reuse these
+            aabbs_ptr         = aabbs;
+            voxel_indices_ptr = voxel_indices;
+        }
+
+        return { aabbs_ptr, voxel_indices_ptr, count };
     }
 
     /// Computes the gradient for a specific gradient
-    Vector3f voxel_grad(const Point3f& p, const Point3i& voxel_index) const {
+    Vector3f voxel_grad(const Point3f &p, const Point3i &voxel_index) const {
```

Comment:
(unrelated to the embree support)
Would this be more efficient?
```
Vector3f voxel_grad(const Point3f& p, const Point3i& voxel_index, Mask active = true) const {
        Float f000, f001, f010, f011, f100, f101, f110, f111;
        dr::Array<Float *, 8> fetch_values;
        fetch_values[0] = &f000;
        fetch_values[1] = &f100;
        fetch_values[2] = &f010;
        fetch_values[3] = &f110;
        fetch_values[4] = &f001;
        fetch_values[5] = &f101;
        fetch_values[6] = &f011;
        fetch_values[7] = &f111;
        m_grid_texture.eval_fetch(rescale_point(p), fetch_values, active);

        Point3f p_local = p * m_resolution - Point3f(voxel_index);
        Float y0 = dr::lerp(f100 - f000, f110 - f010, p_local.y());
        Float y1 = dr::lerp(f101 - f001, f111 - f011, p_local.y());
        Float dx = dr::lerp(y0, y1, p_local.z());
        Float x0 = dr::lerp(f010 - f000, f110 - f100, p_local.x());
        Float x1 = dr::lerp(f011 - f001, f111 - f101, p_local.x());
        Float dy = dr::lerp(x0, x1, p_local.z());
              x0 = dr::lerp(f001 - f000, f101 - f100, p_local.x());
              x1 = dr::lerp(f011 - f010, f111 - f110, p_local.x());
        Float dz = dr::lerp(x0, x1, p_local.y());

        return Vector3f(dx, dy, dz);
    }
```

Reply:
Thanks! I don't think CUDA will do a texture gather for a 3d texture, so I think under the hood it will still be 8 individual lookups. However, imo, on the CPU-side what you've got looks cleaner to me so happy to give it a go.

Reply:
Indeed, `eval_fetch` is the way to go here. I wonder why I didn't do that initially...